pkg/manifests/tls.go

```diff
@@ -36,94 +39,145 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
 	return now().After(latestTime)
 }
 
-// This method creates a central secret containing GRPC TLS key material
-// if the given secret is nil
-// or updates it if the CA present in the given secret is about to expire.
+func (f *Factory) GRPCSecret() (*v1.Secret, error) {
+	s, err := f.NewSecret(MustAssetReader(ClusterMonitoringGrpcTLSSecret))
+	if err != nil {
+		return nil, err
+	}
+
+	s.Namespace = f.namespace
+	s.Data = make(map[string][]byte)
+	s.Annotations = make(map[string]string)
+
+	return s, nil
+}
+
+// RotateGRPCSecret rotates key material for Thanos GRPC TLS based communication.
 //
+// If no key material is present, it creates it.
 // It "rotates" the CA 1/5 before the expiry timespan.
-// The rotation scheme is very naive, it simply creates a new self signed CA
+//
+// The rotation scheme refreshes the expiry of the exising CA
 // and refreshes all the subsequent GRPC server/client certs.
 // For simplicity, the CA as well as all certificates have the same expiration,
 // that is crypto.DefaultCertificateLifetimeInDays = 2 years
 // in order to align rotation.
-func (f *Factory) GRPCSecret(s *v1.Secret) (*v1.Secret, error) {
-	if s == nil {
+func RotateGRPCSecret(s *v1.Secret) error {
+	var (
+		curCA, newCA              *crypto.CA
+		curCABytes, crtPresent    = s.Data["ca.crt"]
+		curCAKeyBytes, keyPresent = s.Data["ca.key"]
+		rotate                    = !crtPresent || !keyPresent
+	)
+
+	if crtPresent && keyPresent {
 		var err error
-		s, err = f.NewSecret(MustAssetReader(ClusterMonitoringGrpcTLSSecret))
+		curCA, err = crypto.GetCAFromBytes(curCABytes, curCAKeyBytes)
 		if err != nil {
-			return nil, err
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
+		} else if needsNewCert(curCA.Config.Certs[0].NotBefore, curCA.Config.Certs[0].NotAfter, time.Now) {
+			rotate = true
+			fmt.Println("##### certs expired")
 		}
-		s.Namespace = f.namespace
-		s.Data = make(map[string][]byte)
-		s.Annotations = make(map[string]string)
 	}
 
-	crt, crtPresent := s.Data["ca.crt"]
-	key, keyPresent := s.Data["ca.key"]
+	if _, ok := s.Annotations["monitoring.openshift.io/grpc-forced-rotate"]; ok {
```

Comment:
```suggestion
	if _, ok := s.Annotations["monitoring.openshift.io/grpc-forced-cert-rotate"]; ok {
```
Let's be a little bit more explicit what it is about

Reply:
it's not just the cert, it is the complete key material, including server/client certs and keys, hence i suggest `grpc-tls-forced-rotate` to gather the whole context.